signature/openpgp/openpgp.go

```diff
@@ -0,0 +1,86 @@
+package openpgp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/containers/image/types"
+
+	"golang.org/x/crypto/openpgp"
+)
+
+type context struct {
+	keyring openpgp.EntityList
+}
+
+type OpenPGPMechanism struct {
+	ctx *context
+}
+
+func NewOpenPGPSigningMechanism() (types.SigningMechanism, error) {
+	return OpenPGPMechanism{ctx: &context{
+		keyring: openpgp.EntityList{},
+	}}, nil
+}
+
+func (m OpenPGPMechanism) ImportKeysFromBytes(blob []byte) ([]string, error) {
+	keyring, err := openpgp.ReadArmoredKeyRing(bytes.NewReader(blob))
+	if err != nil {
+		return nil, err
+	}
+	for _, entity := range keyring {
+		if entity.PrimaryKey == nil {
+			continue
+		}
+		m.ctx.keyring = append(m.ctx.keyring, entity)
+	}
+	if len(m.ctx.keyring) == 0 {
+		return nil, errors.New("no public keys found")
```

Comment:
Strictly speaking, undefined :)

Right now this is not intended as a generic abstraction over key _management_ systems (import/export/list keys etc.), so that someone could build a key management CLI on top. `SigningMechanism` is really here only to sign and verify.

The only place where `ImportKeysFromBytes` is called in production is `prSignedBy`, which creates a completely separate GPG home directory in the gpgme implementation: the idea is that, regardless of anything in $HOME or $GNUPGHOME, `prSignedBy` uses _only_ the keys specified in the `prSignedBy` object.

So, with gpgme, keys are stored in files and keys need to be imported into the keyring before being used; therefore, `ImportKeysFromBytes` has a persistent effect (adds keys to `$optionalDir` or so, and they will persist for future uses of that directory); `prSignedBy` creates a private directory, imports keys exactly once, uses that directory for one (or more, in the future) verifications. 

With openpgp, we can use in-memory keys; so, there is no _need_ for `ImportKeysFromBytes` to have a persistent effect; `prSignedBy` can create an in-memory object (**TODO** we should skip creating the directory in that case, and move that into the `SigningMechanism` as well), import keys into it, and use it for validation.

As it happens, `ImportKeysFromBytes` is always called only once per directory / object / (in the future) context in `prSignedBy`, so it does not matter whether the keyring is reset, hence the “strictly speaking, undefined”.

But, per naming of the function, it seems more reasonable to append keys than to override; it is `Import`, not `OverrideWithImport` after all. And _if_ `SigningMechanism` ever grew into the export/import/list key management abstraction, a non-destructive import seems much more useful to me.

As for duplicate keys (note that they may be duplicate both between old keyring contents and the imported data, and within just the imported data), as long as signatures validate using the imported keys and only the imported keys, we don’t really care _right now_ (well, other than one somewhat misguided test in `mechanism_test.go`, but we can change that one).

Eventually, to implement `SBKeyTypeSignedByGPGKeys`, if that ever happens, we will need to care about signatures attached to public keys, and the ability to import new signatures to already existing public keys, and this will become more complex. But that’s not a constraint right now.